[HUDI-7447] Fix not bootstrap when subTask restart when OPCoordinator handle CheckPointComplete not finished

wenbingshen opened the pull request:
### Change Logs

1. In Insert mode, when the SubTask is restarted, the OperatorCoordinator is in the notifyCheckpointComplete of CheckpointId-100 for a long time. This may be due to the time-consuming processing of some tableService scanning hdfs, or the time-consuming hdfs execution encountered during Rollback and initInstant.
2. At this time, ckp-meta/instantId.INFLIGHT is not completed, but the corresponding commit file has been submitted. At this time, the bootstrap event will be sent when the subTask restarts.
3. After the OperatorCoordinator completes processing the notifyCheckpointComplete, it will create a new Instant, and the subTask will create the corresponding parquet file, etc. based on the Instant.
4. OperatorCoordinator then processes the bootstrap event, creates another new Instant, and rolls back the Instant created in the third step. This causes OperatorCoordinator and Operator to begin to be inconsistent.

This is related to Hudi's three-stage submission, including data snapshot, submit commit file, and submit ckp_meta file

A radical solution to the problem is shown in this PR submission,

Another possibility to solve the problem I am currently experiencing is: in Insert mode, when the task restarts, if the commit has been submitted but ckp_meta has not been submitted, the bootstrap event will not be sent. This can only solve the problem of OPC getting stuck when submitting the ckp_meta file, but when tableService runs and creates new Instant stuck will not be addressed :
```java
protected void sendBootstrapEvent() {
    int attemptId = getRuntimeContext().getAttemptNumber();
    if (attemptId > 0) {
      // either a partial or global failover, reuses the current inflight instant
      if (this.currentInstant != null && !metaClient.getActiveTimeline().filterCompletedInstants().containsInstant(currentInstant)) {
        LOG.info("Recover task[{}] for instant [{}] with attemptId [{}]", taskID, this.currentInstant, attemptId);
        this.currentInstant = null;
        return;
      } else if (this.currentInstant != null) { // <== ****** Add this judgment ******
        LOG.info("Recover task[{}] when coordinator committing for instant [{}] is in progress with attemptId [{}]."
                + " wait for this currentInstant commit complete.", taskID, this.currentInstant, attemptId);
        return;
      }
      // the JM may have also been rebooted, sends the bootstrap event either
    }
    this.eventGateway.sendEventToCoordinator(WriteMetadataEvent.emptyBootstrap(taskID));
    LOG.info("Send bootstrap write metadata event to coordinator, task[{}].", taskID);
  }
```

### Impact

Flink append mode.

### Risk level (write none, low medium or high below)

high

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

wenbingshen commented:
@hudi-bot run azure

danny0405 commented:
> After the OperatorCoordinator completes processing the notifyCheckpointComplete, it will create a new Instant, and the subTask will create the corresponding parquet file, etc. based on the Instant.
OperatorCoordinator then processes the bootstrap event, creates another new Instant, and rolls back the Instant created in the third step. This causes OperatorCoordinator and Operator to begin to be inconsistent.

What is exactly the instant time of the bootstrap instant (should be the inflight instant if the metadata file not committed or empty if it is), and why the coordinator rollsback the instant then if the instant has been committed successfully?

wenbingshen commented:
> What is exactly the instant time of the bootstrap instant (should be the inflight instant if the metadata file not committed or empty if it is), and why the coordinator rollsback the instant then if the instant has been committed successfully?
@danny0405 
1. Hudi StreamWriteOperatorCoordinator has a single-threaded thread pool executor to serially execute some checkpoint and operator events;
2. When notifyCheckpointComplete is executed in the executor, the execution is asynchronous. At this time, Flink will consider that the ck has been executed. Assume that the instant submitted by this ck is Instant A. Because reading HDFS is stuck for a long time, we encounter One situation I encountered is that the scheduleCompaction was stuck for 15 minutes, and the ck timeout was 5 minutes, which exceeded the timeout of flink checkpoint. The status at this time was that the StreamWriteOperatorCoordinator completed the submission of the deltacommit file complete, and the ckp_meta file completed the submission complete. , Flink subTask will time out in the instantToWrite method because the new Instant is not initialized. If the task is configured to retry, the Task will restart;
3. After the Task restarts, because the deltacommit file has been completed and the ckp_meta file has been completed, the uncommitted write metadata event will not be sent to the coordinator, but the bootstrap write metadata event will be sent to the coordinator. At this time, all bootstrap events sent by the task are in the StreamWriteOperatorCoordinator Queued in the executor queue, because the StreamWriteOperatorCoordinator is stuck on scheduleCompaction at this time, the thread worker is still running and cannot handle subsequent bootstrap events;
4. When StreamWriteOperatorCoordinator finishes processing scheduleCompaction and creates new Instant B, subTask knows the creation of new Instant B based on ckp_meta coordination. When subTask processElement, a new parquet file will be created to write data (in Insert mode). The executor of StreamWriteOperatorCoordinator then obtains the bootstrap event in the queue for processing and generates a new Instant C. In startInstant function, it will Rollback Instant B because B is now only in the Inflight state;
5. When C is generated and B is rolled back by the Coordinator, the subTask still uses the parquet file corresponding to B to write. When the next flink checkpoint is executed, the data of B will be snapshotted on the timeline of C.
6. As a result, the StreamWriteOperatorCoordinator and the subTask Operator are incoordinated, violating the three-phase submission protocol in Hudi, as I said above, including data snapshot, submit commit file, and submit ckp_meta file

danny0405 commented:
> When subTask processElement , a new parquet file will be created to write data (in Insert mode). The executor of StreamWriteOperatorCoordinator then obtains the bootstrap event in the queue for processing and generates a new Instant C . In startInstant function, it will
Rollback Instant B  because B is now only in the Inflight state;

Makes sense for me. Insert mode does not support compaction, do you mean `scheduleClustering`?

wenbingshen commented:
> Makes sense for me. Insert mode does not support compaction, do you mean `scheduleClustering`?

We are stuck on scheduleCompaction. Compaction is not supported in Insert mode, but we have not turned it off before. When we encountered problems, we decided to turn it off for optimization, but whether it is scheduleCompaction, scheduleClustering, or other operations, because in There are some other operations that need to be performed after commitInstant. If you encounter some kind of problem during this operation that causes the execution to take a long time, such as reading HDFS, writing HDFS, etc., it will cause the problems I described above.
@danny0405

wenbingshen commented:
Based on the current Hudi three-phase submission protocol: including data snapshot, submit commit file, and submit ckp_meta

If you need to ensure that the `StreamWriteOperatorCoordinator` and `Operator` are coordinated, a better way when the above problem occurs is to let the task backpressure, subTask cannot handle data writing, instead of letting subTask create a parquet file based on Instant B, and then B is C Rolling back creates an incongruous situation, so this PR's workaround can do just that.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteOperatorCoordinator.java:
```diff
@@ -434,9 +434,19 @@ private void handleBootstrapEvent(WriteMetadataEvent event) {
-          .findFirst().map(WriteMetadataEvent::getInstantTime)
-          .orElse(WriteMetadataEvent.BOOTSTRAP_INSTANT);
+              .filter(evt -> evt.getWriteStatuses().size() > 0)
+              .findFirst().map(WriteMetadataEvent::getInstantTime)
```
Unnecessary change.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteOperatorCoordinator.java:
```diff
@@ -434,9 +434,19 @@ private void handleBootstrapEvent(WriteMetadataEvent event) {
+
+      // if currentInstant is pending && bootstrap event instant is empty
+      // reuse currentInstant, reject bootstrap
+      if (this.metaClient.getActiveTimeline().filterInflightsAndRequested().containsInstant(this.instant) && instant
```
We need to reload the active timeline to ensure it is the latest one.

wenbingshen replied:
Addressed. According to the repairs and test cases submitted by previous people, in this case, we only need to reuse the pending instant in `INSERT` mode.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteOperatorCoordinator.java:
```diff
@@ -434,9 +434,19 @@ private void handleBootstrapEvent(WriteMetadataEvent event) {
+      // reuse currentInstant, reject bootstrap
+      if (this.metaClient.getActiveTimeline().filterInflightsAndRequested().containsInstant(this.instant) && instant
+              .equals(WriteMetadataEvent.BOOTSTRAP_INSTANT)) {
+        LOG.info("Reuse current pending Instant {}, ignoring empty bootstrap event.", this.instant);
```
Use `LOG.warn`.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/TestWriteCopyOnWrite.java:
```diff
@@ -169,8 +169,10 @@ public void testPartialFailover() throws Exception {
         .subTaskFails(0, 3)
-        // the last checkpoint instant was rolled back by subTaskFails(0, 2)
-        // with EAGER cleaning strategy
+        // the last checkpoint instant can not rolled back by subTaskFails(0, 2)
```
`can not rolled` -> `can not be rolled`

danny0405 replied:
And while we introduce new changes, also ensure the original test case is covered instead of just modifying the test to pass.

wenbingshen replied:
Agree. I added a new test and gave some reasons for the changes to the original test:
In `INSERT` mode, only job-level restart can roll back the pending Instant. When SubTasks fail to restart, the pending Instant needs to be reused.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/TestWriteCopyOnWrite.java:
```diff
@@ -169,8 +169,10 @@ public void testPartialFailover() throws Exception {
+        // because last data has been snapshot by checkpoint complete but instant has not been commit
+        // so we need re-commit it
+        .assertEmptyEvent()
         .assertNoEvent()
```
Should we remove this line?

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/TestWriteCopyOnWrite.java:
```diff
@@ -665,8 +667,9 @@ public void testRollbackFailedWritesWithLazyCleanPolicy() throws Exception {
         .subTaskFails(0, 1)
-        // the last checkpoint instant was not rolled back by subTaskFails(0, 1)
-        // with LAZY cleaning strategy because clean action could roll back failed writes.
+        // the last checkpoint instant can not rolled back by subTaskFails(0, 1)
```
ditto

wenbingshen commented:
@danny0405 Thanks for review. I have addressed all your comments. PTAL. In response to your question, I responded to your comment and added comments to the code to explain it.

wenbingshen commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/utils/TestFunctionWrapper.java:
```diff
@@ -82,6 +82,14 @@ default void coordinatorFails() throws Exception {
     throw new UnsupportedOperationException();
   }
 
+  /**
+   * Triggers Job level fail, so the coordinator need re-create a new instance.
+   * @throws Exception
+   */
+  default void coordinatorFailsAndRenewOne() throws Exception {
+    throw new UnsupportedOperationException();
+  }
+
```
Triggers Job level fail. During the process of simulating Job restart, we need to reinitialize a new coordinator, so that the coordinator's instant will be initialized to BOOTSTRAP_INSTANT. Otherwise, the old coordinator caches the old instant, which will affect the test results.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/utils/BulkInsertFunctionWrapper.java:
```diff
@@ -160,6 +160,13 @@ public void coordinatorFails() throws Exception {
     this.coordinator.setExecutor(new MockCoordinatorExecutor(coordinatorContext));
   }
 
+  public void coordinatorFailsAndRenewOne() throws Exception {
```
maybe we rename it to `restartCoordinator` ?

wenbingshen replied:
Addressed .PTAL.
